### ISSUE TITLE
$urlMatcherFactoryProvider.type - decoded `null` parameter values are interpreted as `undefined`

### ISSUE THREAD

**Issue by tmcgann** [angular-ui/ui-router#3121]
I created a type of nullable int parameter type (int?) where all non-integer values are treates as `null`. However, when parameter values are decoded and `null` is returned for non-integer values, `transition.params()` and `$transition$.params()` show parameters as `undefined` when I would expect them to be `null`.

Is there a bug with the way `null` parameter values are being interpreted when decoded? Why do I keep seeing `undefined` instead of `null` even when `null` is returned by the decoder?

Plunker: http://plnkr.co/edit/hk6vkZOIJ1R1glxXdXLm?p=preview

**Comment by christopherthielen** [comment]
This is related to array mode handling in the parameter type.  Query parameters are automatically in array mode = "auto".  Array mode creates two "replacements".  [Both the empty string "" and null are mapped to undefined.](https://github.com/ui-router/core/blob/master/src/params/param.ts#L47-L50)

To address this, either turn array mode off or supply your own replace property (an empty array should do).

```js
    params: {
      x: {
        type: 'int?', 
        array: false,
        /* replace: [] // either or */
      }
    },
```

**Comment by christopherthielen** [comment]
This plunker highlights an additional issue.  It appears that during URL synchronization, compares parameters using `==` instead of using the typed parameter's equals function.

**Comment by tmcgann** [comment]
Sincerely sorry to waste your time over a misunderstanding with parameters. Thanks for the guidance though!

While setting `array: false` worked great for me in the plunker, in my actual app, I found that it wasn't enough. I thought it might be because I was performing a transition redirect in an `onBefore` hook, but when I tried to emulate it in the simple plunker example I couldn't get the same behavior I'm seeing in my app: http://plnkr.co/edit/y5QewL6btvbirADBif1Q?p=preview. 

Per your alternate recommendation I ended up using `replace` because I could see when I stepped through the code that `null` kept getting replaced with `''` by the [array mode default replace settings](https://github.com/ui-router/core/blob/master/src/params/param.ts#L47-L50). My replaced ended up looking something like this:

```js
replace: [ { from: null, to: null} ]
```

Your empty array suggestion didn't seem to work for me. I also didn't see it working in the plunker you provided, in that `x: null` didn't show up in the parameters even though the query string updated appropriately. But maybe I was misinterpreting what you said.

Just thought I'd share in case others came across the same problem.

Thanks again for your help!
